File: models/python/models/entities/couchbase/listings.py
from typing import List, Optional, Literal
from clients.couchbase import BaseModelCouchbase, BaseCouchbaseEntityData


class ListingData(BaseCouchbaseEntityData):
    seller_id: str
    registry_name: str
    registry_project_id: Optional[str] = None
    serial_number_range: Optional[str] = None
    project_name: str
    project_type: Literal[
        "afforestation", "renewable", "cookstoves", "methane_capture",
        "fuel_switching", "energy_efficiency", "agriculture", "other"
    ] = "other"
    project_country: Optional[str] = None
    vintage_year: Optional[int] = None
    quantity_tonnes: float = 0.0
    quantity_reserved: float = 0.0
    quantity_sold: float = 0.0
    price_per_tonne_eur: float
    verification_status: Literal["pending", "verified", "failed"] = "pending"
    methodology: Optional[str] = None
    co_benefits: List[str] = []
    description: Optional[str] = None
    supporting_documents: List[str] = []
    status: Literal["draft", "active", "paused", "sold_out"] = "draft"


class Listing(BaseModelCouchbase[ListingData]):
    _collection_name = "listings"

File: clients/typescript/api/listings.ts
import { get, put, del } from './client';

export interface Listing {
    id: string;
    seller_id: string;
    registry_name: string;
    registry_project_id: string | null;
    serial_number_range: string | null;
    project_name: string;
    project_type: string;
    project_country: string | null;
    vintage_year: number | null;
    quantity_tonnes: number;
    quantity_reserved: number;
    quantity_sold: number;
    price_per_tonne_eur: number;
    verification_status: string;
    methodology: string | null;
    co_benefits: string[];
    description: string | null;
    supporting_documents: string[];
    status: string;
}

export interface ListingSearchResponse {
    listings: Listing[];
    count: number;
}

export async function listingsGet(): Promise<ListingSearchResponse> {
    return await get('/listings');
}

export async function listingGetById(id: string): Promise<Listing> {
    return await get(`/listings/${id}`);
}

export async function listingUpdate(id: string, data: Partial<Listing>): Promise<Listing> {
    return await put(`/listings/${id}`, data);
}

export async function listingDelete(id: string): Promise<void> {
    return await del(`/listings/${id}`);
}

export async function listingVerify(id: string): Promise<Listing> {
    return await get(`/listings/${id}/verify`);
}
